### PR TITLE
Calendar Event create/edit/delete

### PR DESCRIPTION
**Pull request by rh12**
refs: [MBL-17293](https://instructure.atlassian.net/browse/MBL-17293)
affects: Student
release note: none

## What's new
- `modify_calendar_event` FF
- Added Add Calendar Event to Calendar screen, changed "+" button to menu
- Added Edit and Delete to Event Details screen, added "..." menu
- Deletion handles repeated events

### Chores
- Some refactors in EditCalendarToDo
- Updated `InstUI.DatePickerCell`
- Added `InstUI.ToggleCell`
- Reverted `InstUI.TextFieldCell` to single line only
- Added `Date.make()` to simplify date creation in tests
- Added `ConfirmationViewModel`, based on `ConfirmationAlert` to display multiple confirm options

## What's not included
- Frequency selection screens and logic is not part of this task
- Edit doesn't handle repeated events, no confirmation dialog is added for now
- Event details editor is not RCE yet

## Test plan
- Smoke test Edit Calendar ToDo
- Test Add / Edit / Delete Calendar Events

## Screenshots

<table>
<tr><th>Before</th><th>After</th></tr>
<tr>
<td></td>
<td></td>
</tr>
</table>

## Checklist
- [ ] Follow-up e2e test ticket created
- [ ] A11y checked
- [ ] Tested on phone
- [ ] Tested on tablet
- [ ] Tested in dark mode
- [ ] Tested in light mode
- [ ] Approve from product

[MBL-17293]: https://instructure.atlassian.net/browse/MBL-17293?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by inst-danger** [comment]
<details><summary>Student Build QR Code:</summary>

</details>

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>Build failed, skipping coverage check</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>This pull request will not generate a release note.</td>
    </tr>
  </tbody>
</table>


#### Affected Apps: Student

[MBL-17293](https://instructure.atlassian.net/browse/MBL-17293)

<details><summary>:x: XCTest failed: CoreTests/BulkPublishLocalStateRefresherTests/testLocalStateRefresh</summary>


    Asynchronous wait failed: Exceeded timeout of 0.1 seconds, with unfulfilled expectations: "Publisher finished".
</details>
<details><summary>:x: XCTest failed: CoreTests/CourseSyncSyllabusInteractorLiveTests/testSuccessfulFetch</summary>


    XCTAssertEqual failed: ("0") is not equal to ("2")
    XCTAssertEqual failed: ("nil") is not equal to ("Optional("1")")
    XCTAssertEqual failed: ("nil") is not equal to ("Optional("2")")
</details>
<details><summary>:x: XCTest failed: CoreTests/PutCalendarEventRequestTests/testProperties</summary>


    XCTAssertEqual failed: ("calendar_events/42") is not equal to ("planner_notes/42")
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 05ea63c97eb049f50fc53f2f9010c583dce91eee
</p>

File: Core/Core/AppEnvironment/Clock.swift [hunk 1]
```diff
@@ -62,4 +62,10 @@ public class Clock {
         #endif
         return Date()
     }
+
+    #if DEBUG
+    public static func date(year: Int, month: Int? = nil, day: Int? = nil, hour: Int? = nil, minute: Int? = nil, second: Int? = nil) -> Date {
+        DateComponents(calendar: .current, year: year, month: month, day: day, hour: hour, minute: minute, second: second).date!
+    }
+    #endif
```
[the remark this answers is not in the document]
Comment (rh12) [reply Core/Core/AppEnvironment/Clock.swift:70]: Good point, thanks!

File: Core/Core/Planner/CalendarEvent/ViewModel/CalendarEventDetailsViewModel.swift [hunk 2]
```diff
@@ -19,26 +19,101 @@
...
+        self.completion = completion
+        self.deleteConfirmation = deleteSingleItemConfirmation
+
+        event
+            .compactMap { $0 }
+            .sink { [weak self] in
+                guard let self else { return }
+                deleteConfirmation = makeDeleteConfirmation(event: $0)
+            }
+            .store(in: &subscriptions)
```
[the remark this answers is not in the document]
Comment (rh12) [reply Core/Core/Planner/CalendarEvent/ViewModel/CalendarEventDetailsViewModel.swift:83]: I'm not sure I fully agree with this one, but I'm extracting these to methods as you described.

**Comment by rh12** [comment]
@vargaat 
> 1. The title input field supports multiple lines. The "Done" button on the keyboard doesn't hide the keyboard but inserts a line break. On iPads instead of "Done" there's a dedicated line break button. On web these new lines render but when editing the page they disappear and cannot be added back. I think this field should be a one liner and the "Done" on the keyboard should just resign the first responder status.

Making it multiline was a UX request, but it causes multiple issues:
- Done button vs newline (as you described)
- Broken display of existing title if it is one line
- The whole multiline SwiftUI TextField implementation (via `axis: .vertical) seems not there yet :/

For now, I've removed multiline support in the component (used both in EditEvent and EditToDo screens)
We could bring this up with Marci, and as an improvement I could imagine to redo the whole component with a wrapped UITextField, and if it's not enough display a multiline Text instead of a TextField when not editing, and keep the actual TextField single line.

> 2. Editing an all day event doesn't restore the edit UI properly. The all day switch is disabled and the event's start and end times become visible and show the same 12 AM value.

Nice catch! It was caused by not populating the `time_zone_edited` field, and you having different timezones set in your account then on device (which is normal, I just didn't run into it). I ended up adding it in this PR.

> 3. This happened one time and I couldn't reproduce it any more: I edited an event where the calendar was set to the current user. I tapped on the calendar selector row to go to the picker. I've seen for a second that the current user is selected then the selection disappeared and no item was selected on the screen. I went back to the previous screen and the calendar row still displayed my user's name. I tapped again on that row and the picker displayed as expected with my user selected.

I neither could reproduce this one, nor find any hint in the code what might cause this...
I'm leaving it for now, but we could keep an eye on it going forward.

**Review by szabinst — APPROVED** [review]
QA + 1

Could change the following below in the next PR.

**Comment by ndrsszsz** [comment]
Found 2 issues:

- When we choose any other day than today (in the example video: tomorrow), and we start to create an event, the default date for the event will still be today. It's not necessarily an issue, but I think it could be misleading, since the iOS calendar works differently (if we navigate to a day and start creating an event there, then the default date will be the day we were standing on when starting the creation).
- When we create an event to another calendar (not the user's own), then it won't be visible after creation. You can see it in the video, that if all calendars are selected, then the event is not visible, but if only the specific calendar is selected, then the event appears.

Let me know if you plan to fix it in this PR or I should create a separate ticket.

https://github.com/user-attachments/assets/f4b0249b-db22-4b83-bb01-38ef30cbd701

File: Core/Core/Planner/CalendarEvent/ViewModel/EditCalendarEventViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+    @Published var startTime: Date?
+    @Published var endTime: Date?
+    @Published var frequencyName: String?
+    @Published var calendarName: String?
+    @Published var location: String = ""
+    @Published var address: String = ""
+    @Published var details: String = ""
+
+    @Published var endTimeErrorMessage: String?
+    @Published var shouldShowSaveError: Bool = false
```
[the remark this answers is not in the document]
Comment (rh12) [reply Core/Core/Planner/CalendarEvent/ViewModel/EditCalendarEventViewModel.swift]: Thanks, I've updated the ones which are not used as bindings.

**Review by ndrsszsz — APPROVED** [review]
QA +1
Created tickets for the previously mentioned 2 issues (Ticket numbers: 17816, 17817).

**Pull request merged**